File: examples/scripts/swift/Sources/Shared/Shared.swift
import Foundation
import ModelHealth

// MARK: - Constants

public let internalActivityNames: Set<String> = ["calibration", "neutral"]

// MARK: - API key

public func loadAPIKey() -> String {
    let args = CommandLine.arguments
    if args.count > 1, !args[1].isEmpty {
        return args[1]
    }
    if let key = dotEnvValue(for: "MODEL_HEALTH_API_KEY") {
        return key
    }
    if let key = ProcessInfo.processInfo.environment["MODEL_HEALTH_API_KEY"] {
        return key
    }
    fputs(
        "Model Health API key not found.\n"
            + "Provide it as a CLI argument or set MODEL_HEALTH_API_KEY in .env or your environment.\n",
        stderr
    )
    exit(1)
}

public func dotEnvValue(for key: String) -> String? {
    let envURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        .appendingPathComponent(".env")
    guard let contents = try? String(contentsOf: envURL, encoding: .utf8) else {
        return nil
    }

    for line in contents.components(separatedBy: .newlines) {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty, !trimmed.hasPrefix("#") else {
            continue
        }

        let parts = trimmed.components(separatedBy: "=")
        guard parts.count >= 2, parts[0].trimmingCharacters(in: .whitespaces) == key else {
            continue
        }

        var value = parts.dropFirst().joined(separator: "=").trimmingCharacters(in: .whitespaces)
        if value.count >= 2,
           (value.hasPrefix("\"") && value.hasSuffix("\"")) || (value.hasPrefix("'") && value.hasSuffix("'"))
        {
            value = String(value.dropFirst().dropLast())
        }
        return value.isEmpty ? nil : value
    }
    return nil
}

// MARK: - File I/O

public var downloadsURL: URL {
    URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
        .appendingPathComponent("downloads")
}

@discardableResult
public func saveFile(named filename: String, data: Data) -> String {
    let dir = downloadsURL
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    let url = dir.appendingPathComponent(filename)
    do {
        try data.write(to: url)
    } catch {
        fputs("Could not save \(filename): \(error)\n", stderr)
    }
    return url.path
}

// MARK: - Polling

public func pollAnalysis(service: ModelHealthService, task: Analysis) async throws -> AnalysisStatus {
    while true {
        let status = try await service.analysisStatus(for: task)
        if case .processing = status {
            print("  Analysing...  ", terminator: "\r")
            fflush(stdout)
        } else {
            print()
            return status
        }
        try await Task.sleep(nanoseconds: 10_000_000_000)
    }
}

// MARK: - Prompts

public func pickOne<T>(from items: [T], prompt: String, label: (T) -> String, defaultIndex: Int? = nil) -> T {
    for (index, item) in items.enumerated() {
        let suffix = index == defaultIndex ? "  (default)" : ""
        print("  \(index + 1). \(label(item))\(suffix)")
    }

    let rangeHint: String
    if let defaultIndex {
        rangeHint = "1–\(items.count), Enter for \(defaultIndex + 1)"
    } else {
        rangeHint = "1–\(items.count)"
    }

    while true {
        print("\n\(prompt) (\(rangeHint)): ", terminator: "")
        let line = readLine()?.trimmingCharacters(in: .whitespaces) ?? ""
        if line.isEmpty, let defaultIndex {
            return items[defaultIndex]
        }
        if let num = Int(line), (1...items.count).contains(num) {
            return items[num - 1]
        }
        print("  Please enter a number between 1 and \(items.count).")
    }
}

public func pickMulti<T>(from items: [T], prompt: String, label: (T) -> String) -> [T] {
    for (index, item) in items.enumerated() {
        print("  \(index + 1). \(label(item))")
    }
    while true {
        print("\n\(prompt) (e.g. 1 2 3): ", terminator: "")
        if let line = readLine() {
            let indices = line.split(separator: " ").compactMap { Int($0) }.map { $0 - 1 }
            let selected = indices.compactMap { (0..<items.count).contains($0) ? items[$0] : nil }
            if !selected.isEmpty {
                return selected
            }
        }
        print("  Please enter one or more numbers between 1 and \(items.count).")
    }
}

public func confirm(_ prompt: String, default defaultValue: Bool? = nil) -> Bool {
    let hint: String
    if defaultValue == true {
        hint = "[Y/n]"
    } else if defaultValue == false {
        hint = "[y/N]"
    } else {
        hint = "[y/n]"
    }

    while true {
        print("\(prompt) \(hint): ", terminator: "")
        let raw = readLine()?.trimmingCharacters(in: .whitespaces).lowercased() ?? ""
        if raw == "y" || raw == "yes" {
            return true
        }

        if raw == "n" || raw == "no" {
            return false
        }

        if raw.isEmpty, let defaultVal = defaultValue {
            return defaultVal
        }

        print("  Please enter y or n.")
    }
}

// MARK: - MotionDataType helpers

public extension MotionDataType {
    var fileExtension: String {
        switch self {
        case .animation: return "json"
        case .kinematics(.mot): return "mot"
        case .kinematics(.csv): return "csv"
        case .markers(.trc): return "trc"
        case .markers(.csv): return "csv"
        case .model: return "osim"
        case .tagged(_, let ext): return ext
        }
    }

    var typeLabel: String {
        switch self {
        case .animation: return "animation"
        case .kinematics(.mot): return "kinematics_mot"
        case .kinematics(.csv): return "kinematics_csv"
        case .markers(.trc): return "markers_trc"
        case .markers(.csv): return "markers_csv"
        case .model: return "model"
        case .tagged(let tag, _): return tag
        }
    }
}

// MARK: - AnalysisDataType helpers

public extension AnalysisDataType {
    var fileExtension: String {
        switch self {
        case .metrics: return "json"
        case .report: return "pdf"
        case .data: return "zip"
        }
    }

    var typeLabel: String {
        switch self {
        case .metrics: return "metrics"
        case .report: return "report"
        case .data: return "data"
        }
    }
}

File: examples/scripts/ts/_shared.ts
/**
 * Shared utilities for Model Health TypeScript example scripts.
 * Mirrors examples/python/_utils.py and _prompts.py.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import type { ModelHealthClient, Analysis, AnalysisStatus } from '@modelhealth/modelhealth';

// MARK: - Constants

export const INTERNAL_ACTIVITY_NAMES = new Set(['calibration', 'neutral']);

export const MOTION_DATA_EXT: Record<string, string> = {
  animation:      'json',
  kinematics_mot: 'mot',
  kinematics_csv: 'csv',
  markers_trc:    'trc',
  markers_csv:    'csv',
  model:          'osim',
};

export const ANALYSIS_DATA_EXT: Record<string, string> = {
  report:  'pdf',
  data:    'zip',
};

// MARK: - API key

export function loadApiKey(cliArg?: string): string {
  if (cliArg) return cliArg;
  const fromEnvFile = dotEnvValue('MODEL_HEALTH_API_KEY');
  if (fromEnvFile) return fromEnvFile;
  const fromEnv = process.env.MODEL_HEALTH_API_KEY;
  if (fromEnv) return fromEnv;
  console.error(
    'Model Health API key not found.\n' +
    'Provide it as a CLI argument or set MODEL_HEALTH_API_KEY in .env or your environment.'
  );
  process.exit(1);
}

export function dotEnvValue(key: string): string | undefined {
  const envPath = path.join(process.cwd(), '.env');
  if (!fs.existsSync(envPath)) return undefined;
  for (const line of fs.readFileSync(envPath, 'utf-8').split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
    const eqIdx = trimmed.indexOf('=');
    if (trimmed.slice(0, eqIdx).trim() !== key) continue;
    let value = trimmed.slice(eqIdx + 1).trim();
    if (
      value.length >= 2 &&
      ((value.startsWith('"') && value.endsWith('"')) ||
       (value.startsWith("'") && value.endsWith("'")))
    ) {
      value = value.slice(1, -1);
    }
    return value || undefined;
  }
  return undefined;
}

// MARK: - File I/O

export const downloadsDir = path.join(process.cwd(), 'downloads');

export function saveFile(filename: string, data: Uint8Array | Buffer): string {
  fs.mkdirSync(downloadsDir, { recursive: true });
  const filePath = path.join(downloadsDir, filename);
  fs.writeFileSync(filePath, data);
  return filePath;
}

// MARK: - Polling

export async function pollAnalysis(
  client: ModelHealthClient,
  task: Analysis,
  intervalMs = 10_000
): Promise<AnalysisStatus> {
  while (true) {
    const status = await client.analysisStatus(task);
    if (status.type === 'processing') {
      process.stdout.write('  Analysing...  \r');
    } else {
      process.stdout.write('\n');
      return status;
    }
    await sleep(intervalMs);
  }
}

export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// MARK: - Prompts

let _rl: readline.Interface | null = null;

function getRL(): readline.Interface {
  if (!_rl) {
    _rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  }
  return _rl;
}

export function prompt(question: string): Promise<string> {
  return new Promise(resolve => getRL().question(question, resolve));
}

export function closePrompts(): void {
  _rl?.close();
  _rl = null;
}

export async function pickOne<T>(
  items: T[],
  question: string,
  label: (item: T) => string,
  defaultItem?: T
): Promise<T> {
  const defaultIdx = defaultItem !== undefined ? items.indexOf(defaultItem) : -1;
  items.forEach((item, i) => {
    const suffix = i === defaultIdx ? '  (default)' : '';
    console.log(`  ${i + 1}. ${label(item)}${suffix}`);
  });
  const rangeHint = defaultIdx >= 0 ? `1–${items.length}, Enter for ${defaultIdx + 1}` : `1–${items.length}`;
  while (true) {
    const raw = (await prompt(`\n${question} (${rangeHint}): `)).trim();
    if (raw === '' && defaultIdx >= 0) return items[defaultIdx];
    const n = parseInt(raw, 10);
    if (!isNaN(n) && n >= 1 && n <= items.length) return items[n - 1];
    console.log(`  Please enter a number between 1 and ${items.length}.`);
  }
}

export async function pickMulti<T>(
  items: T[],
  question: string,
  label: (item: T) => string
): Promise<T[]> {
  items.forEach((item, i) => console.log(`  ${i + 1}. ${label(item)}`));
  while (true) {
    const raw = await prompt(`\n${question} (e.g. 1 2 3): `);
    const indices = raw.trim().split(/\s+/).map(s => parseInt(s, 10) - 1);
    const selected = indices.filter(i => i >= 0 && i < items.length).map(i => items[i]);
    if (selected.length > 0) return selected;
    console.log(`  Please enter one or more numbers between 1 and ${items.length}.`);
  }
}

export async function confirm(question: string, defaultValue?: boolean): Promise<boolean> {
  const hint = defaultValue === true ? '[Y/n]' : defaultValue === false ? '[y/N]' : '[y/n]';
  while (true) {
    const raw = (await prompt(`${question} ${hint}: `)).trim().toLowerCase();
    if (raw === 'y' || raw === 'yes') return true;
    if (raw === 'n' || raw === 'no') return false;
    if (raw === '' && defaultValue !== undefined) return defaultValue;
    console.log('  Please enter y or n.');
  }
}
